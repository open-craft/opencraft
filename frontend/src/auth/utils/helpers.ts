import { V2Api } from 'global/api';

export const checkAuthAndRefreshToken = async () => {
  const refreshToken = localStorage.getItem('token_refresh');
  let response = null;

  if (refreshToken) {
    try {
      response = await V2Api.authRefreshCreate({
        data: { refresh: refreshToken }
      });
    } catch (e) {
      return false;
    }
  }

  if (response) {
    localStorage.setItem('token_access', response.access);
    return true;
  }

  return false;
};
