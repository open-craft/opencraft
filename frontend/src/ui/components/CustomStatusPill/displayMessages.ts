const messages = {
  cancelRedeployment: {
    defaultMessage: 'Cancel deployment',
    description: 'Cancel deployment button.'
  },
  unavailable: {
    defaultMessage: 'Status: loading',
    description: 'Redeployment status.'
  },
  unavailableTooltip: {
    defaultMessage:
      'Instance status is being loaded, please wait a little bit.',
    description: 'Redeployment status tooltip text.'
  },
  runningDeployment: {
    defaultMessage: 'Status: deploying changes',
    description: 'Redeployment status.'
  },
  runningDeploymentTooltip: {
    defaultMessage:
      'Your instance is being updated with the latest settings. ' +
      'If you cancel this deployment, your changes wont be lost, ' +
      'but they will need to be redeployed.',
    description: 'Redeployment status tooltip text.'
  },
  updatedDeployment: {
    defaultMessage: 'Status: Up to date',
    description: 'Redeployment status.'
  },
  updatedDeploymentTooltip: {
    defaultMessage: 'Your instance is being updated with the latest settings.',
    description: 'Redeployment status tooltip text.'
  },
  preparingInstance: {
    defaultMessage: "Status: We're preparing your instance ",
    description: 'Redeployment status.'
  },
  preparingInstanceTooltip: {
    defaultMessage:
      'Your instance is being set-up, come back in a few hours to customize it.',
    description: 'Redeployment status tooltip text.'
  },
  pendingChanges: {
    defaultMessage: 'Status: Pending changes',
    description: 'Redeployment status.'
  },
  pendingChangesTooltip: {
    defaultMessage:
      "Your instance is up, but the latest settings aren't deployed yet.",
    description: 'Redeployment status tooltip text.'
  },
  failed: {
    defaultMessage: 'Status: Deployment error',
    description: 'Redeployment status.'
  },
  failedTooltip: {
    defaultMessage:
      'Your deployment failed. Our on-duty staff has been informed, and is ' +
      'starting to work on fixing it.',
    description: 'Redeployment status tooltip text.'
  }
};

export default messages;
