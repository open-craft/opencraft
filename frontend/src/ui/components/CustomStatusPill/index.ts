export * from './CustomStatusPill';
export * from './DeploymentIndicator';
