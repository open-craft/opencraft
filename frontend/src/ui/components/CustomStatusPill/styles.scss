@import '~styles/theme';

.status-pill {
  background-color: $primary-2;
  border-radius: 39px;
  margin: 4px 2px;
  padding: 8px 10px;
  display: inline-flex;

  .text {
    font-family: $roboto-font;
    text-align: center;
    font-weight: 500;
    font-size: 16px;
    margin-top: 6px;
    margin-left: 10px;
    margin-right: 10px;
  }

  .dot {
    height: 25px;
    width: 25px;
    border: solid 2px #1e4c59;
    border-radius: 50%;
    display: inline-block;
  }

  .cancel-deployment {
    margin-left: 20px;
    margin-right: 10px;

    &:hover {
      text-decoration: underline;
      cursor: pointer;
    }

    i {
      margin-right: 3px;
    }
  }
}
