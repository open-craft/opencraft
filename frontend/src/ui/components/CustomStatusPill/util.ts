import {
  OpenEdXInstanceDeploymentStatusStatusEnum as DeploymentStatus,
  OpenEdXInstanceDeploymentStatusDeploymentTypeEnum as DeploymentType
} from 'ocim-client';

interface StatusContext {
  tooltipText: string;
  pillColor: string;
  deploymentStatusText: string;
}

/**
 * An utility function to conditionally return color, tooltip and
 * status text keys based on deployment status and type.
 *
 * @param redeploymentStatus
 * @param deploymentType
 * @returns StatusContext
 */
export function buildStatusContext(
  redeploymentStatus?: string | null,
  deploymentType?: string | null
): StatusContext {
  const result: StatusContext = {
    pillColor: 'grey',
    deploymentStatusText: 'unavailable',
    tooltipText: 'unavailableTooltip'
  };

  switch (redeploymentStatus) {
    case DeploymentStatus.Healthy:
      result.pillColor = '#00a556';
      result.deploymentStatusText = 'updatedDeployment';
      result.tooltipText = 'updatedDeploymentTooltip';
      break;
    case DeploymentStatus.Provisioning:
      result.pillColor = '#ff9b04';
      // If there's a deployment provisioning, but it's the
      // first on (from registration), show preparing instance
      // message.
      if (deploymentType === DeploymentType.Registration) {
        result.deploymentStatusText = 'preparingInstance';
        result.tooltipText = 'preparingInstanceTooltip';
      }

      // If not, then this is a normal deployment, so show the usual
      // running deployment message.
      else {
        result.deploymentStatusText = 'runningDeployment';
        result.tooltipText = 'runningDeploymentTooltip';
      }
      break;
    case DeploymentStatus.Preparing:
      result.pillColor = '#ff9b04';
      result.deploymentStatusText = 'preparingInstance';
      result.tooltipText = 'preparingInstanceTooltip';
      break;
    case DeploymentStatus.ChangesPending:
      result.pillColor = '#00a556';
      result.deploymentStatusText = 'pendingChanges';
      result.tooltipText = 'pendingChangesTooltip';
      break;
    default:
      // Default values already set up when instancing variables
      break;
  }
  return result;
}
