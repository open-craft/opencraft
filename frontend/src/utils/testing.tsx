import { createMemoryHistory } from 'history';
import * as React from 'react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import * as renderer from 'react-test-renderer';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { createRootReducer } from '../global/reducers';

export const setupComponentForTesting = (
  reactContent: JSX.Element,
  storeContents = {}
) => {
  const middleware = applyMiddleware(thunk);

  const store = createStore(
    createRootReducer(createMemoryHistory()),
    storeContents,
    middleware
  );

  return renderer.create(
    <IntlProvider textComponent={React.Fragment} locale="en">
      <Provider store={store}>
        <MemoryRouter>{reactContent}</MemoryRouter>
      </Provider>
    </IntlProvider>
  );
};
