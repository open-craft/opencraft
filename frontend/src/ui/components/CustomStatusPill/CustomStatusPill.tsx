import * as React from 'react';
import { WrappedMessage } from 'utils/intl';
import { Tooltip, OverlayTrigger, Badge, Nav } from 'react-bootstrap';
import { OpenEdXInstanceDeploymentStatusStatusEnum as DeploymentStatus} from 'ocim-client';

import messages from './displayMessages';
import './styles.scss';

interface Props {
  loading: boolean;
  redeploymentStatus: string | null;
  cancelRedeployment: Function;
}

export const CustomStatusPill: React.FC<Props> = ({loading, redeploymentStatus, cancelRedeployment}) => {
  let dotColor = 'grey';
  let deploymentStatusText = 'unavailable';
  let tooltipText = 'unavailableTooltip';

  switch (redeploymentStatus) {
    case DeploymentStatus.Healthy:
      dotColor = '#1abb64';
      deploymentStatusText = 'updatedDeployment';
      tooltipText = 'updatedDeploymentTooltip';
      break;
    case DeploymentStatus.Provisioning:
      dotColor = '#ff9b04';
      deploymentStatusText = 'runningDeployment';
      tooltipText = 'runningDeploymentTooltip';
      break;
    case DeploymentStatus.Preparing:
      dotColor = '#ff9b04';
      deploymentStatusText = 'preparingInstance';
      tooltipText = 'preparingInstanceTooltip';
      break;
    case DeploymentStatus.ChangesPending:
      dotColor = '#1abb64';
      deploymentStatusText = 'pendingChanges';
      tooltipText = 'pendingChangesTooltip';
      break;
    default:
      // Default values already set up when instancing variables
      break;
  }

  const tooltip = (
    <Tooltip id="redeployment-status">
      <WrappedMessage messages={messages} id={tooltipText} />
    </Tooltip>
  );

  return (
    <OverlayTrigger placement="right" overlay={tooltip}>
      <Badge pill className="status-pill" variant="primary">
        <span className="dot" style={{ backgroundColor: dotColor }} />
        <div className="text">
          <WrappedMessage id={deploymentStatusText} messages={messages} />
        </div>
        {redeploymentStatus === DeploymentStatus.Provisioning &&
          !loading && (
            <Nav
              className="text cancel-deployment"
              onClick={() => {
                cancelRedeployment();
              }}
            >
              <i className="fas fa-xs fa-times" />
              <WrappedMessage id="cancelRedeployment" messages={messages} />
            </Nav>
          )}
      </Badge>
    </OverlayTrigger>
  );
};
