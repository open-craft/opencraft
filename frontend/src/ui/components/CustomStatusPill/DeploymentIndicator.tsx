import * as React from 'react';
import { buildStatusContext } from './util';

interface Props {
  redeploymentStatus?: string | null;
  deploymentType?: string | null;
}

export const DeploymentIndicator: React.FC<Props> = (props: Props) => {
  const { redeploymentStatus, deploymentType } = props;
  const { pillColor } = buildStatusContext(redeploymentStatus, deploymentType);
  const style = {
    color: pillColor,
    fontSize: '12px'
  };
  return <i className="fa fa-circle" style={style} />;
};
