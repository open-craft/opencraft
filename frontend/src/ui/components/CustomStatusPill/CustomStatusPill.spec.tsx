import React from 'react';
import { setupComponentForTesting } from "utils/testing";
import { CustomStatusPill } from './CustomStatusPill';

it('renders without crashing', () => {
    const tree = setupComponentForTesting(<CustomStatusPill />).toJSON();
    expect(tree).toMatchSnapshot();
});
