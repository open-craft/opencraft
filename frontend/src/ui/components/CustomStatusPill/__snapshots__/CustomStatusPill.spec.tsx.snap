// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<span
  className="status-pill badge badge-pill badge-primary"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <span
    className="dot"
    style={
      Object {
        "backgroundColor": "grey",
      }
    }
  />
  <div
    className="text"
  >
    Status: loading
  </div>
</span>
`;
